src/Documentation/Event-Sourcing/GrainStateAPI.md

```diff
@@ -0,0 +1,112 @@
+
+The runtime updates the grain state _automatically_ whenever events are raised. There is no need for the application to explicitly update the state after raising an event. However, the application still has to provide the code that specifies _how_ to update the state in response to an event. This can be done in two ways.
+
+**(a)** The GrainState class can implement one or more `Apply` methods on the `StateType`. Typically, one would create multiple overloads, and the closest match is chosen for the runtime type of the event:
+```csharp
+class GrainState {
+   
+   Apply(E1 @event)  
+   {
+     // code that updates the state
+   }
+   Apply(E2 @event)  
+   {
+     // code that updates the state
+   }
+}
+```
+
+**(b)** The grain can override the TransitionState function:
+```csharp
+protected override void TransitionState(State state, EventType @event)
+{
+   // code that updates the state
+}
+```
+
+The transition methods are assumed to have no side effects other than modifying the state object, and should be deterministic (otherwise, the effects are unpredictable).  If the transition code throws an exception, that exception is caught and included in a warning in the Orleans log, issued by the log-consistency provider.  
+
+When, exactly, the runtime calls the transition methods depends on the chosen log consistency provider and its configuration. It is best for applications not to rely on a particular timing, except when specifically guaranteed by the log consistency provider. 
+
+Some providers, such as the `LogStorage` log-consistency provider, replay the event sequence every time the grain is loaded. Therefore, as long as the event objects can still be properly deserialized from storage, it is possibly to radically modify the GrainState class and the transition methods. But for other providers, such as the `StateStorage` log-consistency provider, only the `GrainState` object is persisted, so developers must ensure that it can be deserialized correctly when read from storage. 
+
+
+## Raising Multiple Events
+
+It is possible to make multiple calls to RaiseEvent before calling ConfirmEvents:
+
+```csharp
+RaiseEvent(e1);
+RaiseEvent(e2);
+await ConfirmEvents();
+```
+
+However, this is likely to cause two successive storage accesses, and it incurs a risk that the grain fails after writing only the first event. Thus, it is usually better to raise multiple events at once, using
+
+```csharp
+RaiseEvents(IEnumerable<EventType> events)
+```
+
+This guarantees that the given sequence of events is written to storage atomically. Note that since the version number always matches the length of the event sequence, raising multiple events increases the version number by more than one at a time.
+
+
+## Retrieving the Event Sequence
+
+The following method allows the application to retrieve a specified segment of the sequence of all confirmed events:
+
+```csharp
+Task<IReadOnlyList<EventType>> RetrieveConfirmedEvents(int fromVersion, int toVersion)
```

Comment:
Will this method have  version with datetime or timespans or another timebased parameters?
like
```csharp
Task<IReadOnlyList<EventType>> RetrieveConfirmedEvents(DateTime fromVersion, DateTime toVersion)
```